pokedex/simulator/pokedex/src/simulator.rs

```diff
@@ -359,3 +335,314 @@ impl Drop for SimulatorHandle {
+                        boxed,
+                    ))
+                }
+            }
+        }
+
+        let mut overlapped_segment = None;
+        let mut unchecked_index: Vec<Range<u32>> =
+            segments.iter().map(|(index, _)| index.clone()).collect();
+        unchecked_index.sort_by_key(|range| range.start);
+        for window in unchecked_index.windows(2) {
+            let addr1 = &window[0];
+            let addr2 = &window[1];
+            if addr2.start < addr1.end {
+                overlapped_segment = Some(&window[1]);
+            }
+        }
+
+        if let Some(range) = overlapped_segment {
+            miette::bail!(
+                "Address space with offset={:#010x} length={:#010x} overlapped previous address space",
+                range.start,
+                range.end - range.start
+            )
+        }
+
+        Ok(Self {
+            interrupt_controllers: ic.unwrap(),
+            bus_bridge: BusBridge {
+                address_space: segments,
+            },
+        })
+    }
+}
+
+#[test]
+fn test_address_space() -> miette::Result<()> {
+    let _ = BusInfo::from_device_config("./assets/devices.kdl")?;
+    Ok(())
+}
+
+#[derive(Debug)]
+pub enum BusError {
+    OutOfMemory {
+        id: &'static str,
+        addr: u32,
+        len: u32,
+    },
+    Illegal,
+}
+
+pub trait Addressable: std::fmt::Debug {
+    fn resp_bus_read(&self, offset: u32, dest: &mut [u8]) -> Result<(), BusError>;
+    fn resp_bus_write(&mut self, offset: u32, data: &[u8]) -> Result<(), BusError>;
+}
+
+#[derive(Debug)]
+pub struct NaiveMemory {
+    memory: Vec<u8>,
+}
+
+impl NaiveMemory {
+    pub fn new(size: usize) -> Self {
+        Self {
+            memory: vec![0u8; size],
+        }
+    }
+}
+
+impl Addressable for NaiveMemory {
+    /// read return a slice of the inner memory. Caller should guarantee index and read length is
+    /// valid. An out of range slicing will directly bail out.
+    fn resp_bus_read(&self, offset: u32, dest: &mut [u8]) -> Result<(), BusError> {
+        let length = self.memory.len() as u32;
+        if offset >= length || offset + dest.len() as u32 > length {
+            return Err(BusError::OutOfMemory {
+                id: "NaiveMemoryRead",
+                addr: offset,
+                len: dest.len() as u32,
+            });
+        }
+
+        dest.copy_from_slice(&self.memory[offset as usize..offset as usize + dest.len()]);
+
+        Ok(())
+    }
+
+    fn resp_bus_write(&mut self, offset: u32, data: &[u8]) -> Result<(), BusError> {
+        let length = self.memory.len() as u32;
+        if offset >= length || offset + data.len() as u32 > length {
+            return Err(BusError::OutOfMemory {
+                id: "NaiveMemoryWrite",
+                addr: offset,
+                len: data.len() as u32,
+            });
+        }
+
+        (&mut self.memory[offset as usize..offset as usize + data.len()]).copy_from_slice(&data);
```

Comment:
no need to `&mut` manually here. If you are not so familiar with Rust, you can use `cargo clippy`, it will give you some useful suggestions.